docs/src/writing_good_rules.md

```diff
@@ -96,7 +96,7 @@ function rrule(::typeof(*), A::AbstractMatrix, B::AbstractMatrix)
     function times_pullback(ȳ)
         dA = ȳ * B'
         dB = A' * ȳ
-        return NoTangent(), project_A(dA), project_B(dB)
+        return NoTangent(), @thunk(project_A(dA)), @thunk(project_B(dB))
```

Comment:
I think this adds complexity to this example that is not warranted.
The code is correct and illustrates the above point correctly.

It is an example,  not the canonical, optimized implementation.

Reply:
It's also backwards, I think? Presumably most of the work is in `*` which is outside the thunk.

Reply:
This is not the place to document this.
Here we just want to make it clear how `ProjectTo` is used.

We might like to add a 
```
!!! Note "It is often good to @thunk your projections"
    The above example is potentially good place for using an [`@thunk`](@ref).
    This is not required but can be usedful ... see [docs on thunking](<link>)...
    ...

    `project` itself will when applied to `AbstractThunk`s make a deferred operation, rather than unthunking directly.
    But this is not the case for most operations such as `*`
    ...
    A more optimized implementation of the above would have
    ```julia
    times_pullback(ȳ) = NoTangent(), @thunk(project_A(ȳ * B')), @thunk(project_B(A' * ȳ))
    ```

    Thus deferring both the application of the product rule, and the projection until(/if) the tangent gets used.
```

Reply:
> It's also backwards, I think? Presumably most of the work is in * which is outside the thunk.

yes, good catch, should've paid more attention!

Reply:
@oxinabox thanks for the write-up - in general that's exactly what I think was missing in the docs at that point.
Can you just explain this part:

> `project` itself will when applied to `AbstractThunk`s make a deferred operation, rather than unthunking directly. But this is not the case for most operations such as `*`

which still isn't entirely clear to me